lib/Frontend/Frontend.cpp

```diff
@@ -557,24 +698,26 @@ void CompilerInstance::performSema() {
                             options.WarnLongFunctionBodies,
                             options.WarnLongExpressionTypeChecking,
                             options.SolverExpressionTimeThreshold);
+}
 
-  // Even if there were no source files, we should still record known
-  // protocols.
-  if (auto *stdlib = Context->getStdlibModule())
-    Context->recordKnownProtocols(stdlib);
-
-  if (DelayedCB) {
-    performDelayedParsing(MainModule, PersistentState,
-                          Invocation.getCodeCompletionFactory());
-  }
-
-  // Perform whole-module type checking.
+void CompilerInstance::typeCheckMainModule(
+    OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  SharedTimer timer("performSema-parseMainAndTypeCheckTopLevelFiles-typeCheckMainModule");
   if (TypeCheckOptions & TypeCheckingFlags::DelayWholeModuleChecking) {
-    for (auto File : MainModule->getFiles())
-      if (auto SF = dyn_cast<SourceFile>(File))
-        performWholeModuleTypeChecking(*SF);
+    performWholeModuleTypeCheckingOnMainModule();
   }
+  finishTypeCheckingMainModule();
+}
+
+void CompilerInstance::performWholeModuleTypeCheckingOnMainModule() {
```

Comment:
In this case "on main module" is redundant; there's nothing else you can perform whole-module type checking on.

Reply:
Since performWholeModuleTypeChecking was already used, I merged this routine and its sibling into its caller.